### PR TITLE
GEODE-7851: Pulse refreshes expired access tokens

### PR DESCRIPTION
**Pull request by demery-pivotal**
If a user's access token expires, Pulse attempts to refresh it. If the
refresh fails, Pulse logs the user out and redirects the browser to
/pulse/clusterLogout.

Changes in Repository:
- When OAuth is configured, before returning the user's cluster,
  getCluster() checks whether the user's access token has expired.
- If the access token has expired, the repository attempts to refresh
  it.  If the refresh succeeds, the repository reconnects the user's
  cluster to JMX and returns it.
- If the refresh fails, the repository disconnects the user's cluster
  from JMX, removes the cluster from the repository, and throws an
  authentication or authorization exception.

Changes in PulseController:
- If the service call throws an authentication or authorization
  exception, PulseController.  getPulseUpdate() returns a 401 status.

Changes in pulsescript/common.js:
- If a Pulse ajax call returns a 401 status, ajaxPost() redirects the
  browser to /pulse/clusterLogout to log the user out and request
  re-authorization.

Co-authored-by: Joris Melchior <joris.melchior@gmail.com>
Co-authored-by: Dale Emery <demery@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java [hunk 1]
```diff
@@ -80,8 +88,95 @@ public Repository(OAuth2AuthorizedClientService authorizedClientService,
...
+  /**
+   * Returns the cluster for the user associated with the given authentication. If the user's
+   * access token is expired, it is refreshed and the cluster is reconnected to JMX using the fresh
+   * token. If the refresh fails, the user's cluster is disconnected from JMX and removed from the
+   * repository.
+   */
+  private Cluster getClusterWithAuthenticationToken(OAuth2AuthenticationToken authentication) {
+    OAuth2AuthorizedClient authorizedClient = getAuthorizedClient(authentication);
+
+    if (isExpired(authorizedClient.getAccessToken())) {
```
Comment (jinmeiliao) [inline geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java:142]: wondering which flow is better. Using the following might get rid of the reconnectToGemfire method in `Cluster`.
```suggestion
    String userName = authorizedClient.getPrincipalName();

    if (isExpired(authorizedClient.getAccessToken())) {
      logoutUser(userName);
      authorizedClient = refreshExpiredClient(authentication, authorizedClient);
    }

    userName = authorizedClient.getPrincipalName();
    String credentials = authorizedClient.getAccessToken().getTokenValue();
    return getClusterWithCredentials(userName, credentials);
```

Comment (demery-pivotal) [reply geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java:142]: `logoutUser()` discards any data cached in the `Cluster`, including all of the trends stored in circular buffers. That seems like a harsh thing to do when the token refreshes.

File: geode-pulse/build.gradle [hunk 2]
```diff
@@ -50,7 +50,9 @@ dependencies {
   // Needed to fully use log4j instead of commons-logging.
   implementation('org.apache.logging.log4j:log4j-jcl')
   implementation('org.apache.logging.log4j:log4j-api')
-//  implementation('org.apache.logging.log4j:log4j-core')
+  implementation('org.apache.logging.log4j:log4j-slf4j-impl') {
```
Comment (jinmeiliao) [inline geode-pulse/build.gradle:53]: probably it would be good to do this in a separate PR so that we can backport this to support branches.

Comment (demery-pivotal) [reply geode-pulse/build.gradle:53]: Okay. And probably it should be `runtimeOnly` instead of `implementation`

File: geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java [hunk 3]
```diff
@@ -80,8 +88,95 @@ public Repository(OAuth2AuthorizedClientService authorizedClientService,
...
+
+    return getClusterWithUserNameAndPassword(authentication.getName(), null);
+  }
+
+  public Cluster getClusterWithUserNameAndPassword(String userName, String password) {
+    String[] credentials = {userName, password};
+    return getClusterWithCredentials(userName, credentials);
+  }
+
+  public Cluster getClusterWithCredentials(String userName, Object credentials) {
```
[the remark this answers is not in the document]
Comment (demery-pivotal) [reply geode-pulse/src/main/java/org/apache/geode/tools/pulse/internal/data/Repository.java:117]: Done.

File: geode-pulse/build.gradle (hunk 2 above)
Comment (demery-pivotal) [reply geode-pulse/build.gradle]: Done: https://github.com/apache/geode/pull/4988

And I rebased this PR on that one.
